dss-collection-nft/contracts/DSSCollection.cdc

```diff
@@ -0,0 +1,526 @@
+        pub let startTime: UFix64?
+        pub let endTime: UFix64?
+        pub let timeBound: Bool
+        pub var editionIDInCollectionGroup: {UInt64: Bool}
+
+        pub fun editionIDExistsInCollectionGroup(editionID: UInt64): Bool {
+           return self.editionIDInCollectionGroup.containsKey(editionID)
+        }
+
+        init (id: UInt64) {
+            if let collectionGroup = &DSSCollection.collectionGroupByID[id] as &DSSCollection.CollectionGroup? {
+                self.id = collectionGroup.id
+                self.name = collectionGroup.name
+                self.productPublicPath = collectionGroup.productPublicPath
+                self.open = collectionGroup.open
+                self.startTime = collectionGroup.startTime
+                self.endTime = collectionGroup.endTime
+                self.timeBound = collectionGroup.timeBound
+                self.editionIDInCollectionGroup = collectionGroup.editionIDInCollectionGroup
+            } else {
+                panic("collectionGroup does not exist")
+            }
+        }
+    }
+
+    // A top-level CollectionGroup with a unique ID and name
+    //
+    pub resource CollectionGroup {
+        pub let id: UInt64
+        pub let name: String
+        pub let productPublicPath: PublicPath
+        pub var open: Bool
+        pub let startTime: UFix64?
+        pub let endTime: UFix64?
+        pub let timeBound: Bool
+        pub var numMinted: UInt64
+        pub var editionIDInCollectionGroup: {UInt64: Bool}
+
+        // Close this collection group
+        //
+        access(contract) fun close() {
+            pre {
+                self.open == true: "not open"
+            }
+
+            self.open = false
+
+            emit CollectionGroupClosed(id: self.id)
+        }
+
+        // Add editionID to collection group
+        //
+        access(contract) fun addEditionToCollectionGroup(editionID: UInt64) {
+            pre {
+                self.open == true: "not open"
+            }
+
+            self.editionIDInCollectionGroup[editionID] = true
+
+            emit EditionAddedToCollectionGroup(editionID: editionID, collectionGroupID: self.id)
+        }
+
+        // Mint a DSSCollection NFT in this group
+        //
+        pub fun mint(completedBy: String, level: UInt64): @DSSCollection.NFT {
+            pre {
+                self.open != true: "cannot mint an open collection group"
+                DSSCollection.validateTimeRange(
+                    timeBound: self.timeBound,
+                    startTime: self.startTime,
+                    endTime: self.endTime
+                ) == true : "cannot mint a collection group outside of time bounds"
+                level <= 10: "token level must be less than 10"
+            }
+
+            // Create the DSSCollection NFT, filled out with our information
+            //
+            let dssCollectionNFT <- create NFT(
+                id: DSSCollection.totalSupply + 1,
```

Comment:
we should use `self.uuid` the resource uuid. Totally unique and we don't have to worry about keeping it unique

Reply:
actually this will be set in resource `NFT` init method

Reply:
Yes, I changed all the id's to uuid. Thanks for the tip, did not know about that.